---
title: "Chemogram Pipeline"
author: "Kristi Lin-Rahardja"
date: "2024-02-22"
output:
  rmdformats::downcute:
    self_contained: true
    number_sections: true
    code_folding: show
    df_print: paged
  pdf_document:
    toc: yes
---

This pipeline should be run after `setup.Rmd` and can be used in place of the remaining scripts. The chemogram can use any number of signatures to rank drug response against the GDSC cell lines.

The signatures used here were extracted using the [CisSig](https://github.com/jessicascarborough/cissig) pipeline published by [Scarborough et al](https://www.nature.com/articles/s41698-023-00375-y.pdf). The parameters used for each signature were the same as CisSig, with the exception of the irinotecan, topotecan, and vorinostat signatures, which used a differential expression cutoff of 0.15 instead of 0.2.


The only manual parts should be listing the full drug names at the end of the `signatures` chunk below.

# Load Packages

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, message=FALSE}
# devtools::install_github("cancerrxgene/gdscIC50", build_vignettes=FALSE) #restart R after downloading

library(here)
library(gdscIC50) #for cleaning and normalizing the raw drug response data
library(pheatmap) #for plotting survival and sig scores
library(gtools) #for generating the permutations/scoring table
library(plyr) #for match_df to get acc score from rank order
library(patchwork) #for plotting
library(tidyverse)

conflicted::conflict_prefer("here", "here")
conflicted::conflict_prefer("filter", "dplyr")
conflicted::conflict_prefer("mutate", "dplyr")
conflicted::conflict_prefer("select", "dplyr")
```

# Load Data

```{r load_prepped_data}
cell_lines = readRDS(here("Data", "cleaned_cell_line_meta.rds"))
subtype_key = readRDS(here("Data", "subtype_key.rds"))
exprs_norm = readRDS(here("Data", "normalized_exprs.rds")) #exprs data z-scores
# exprs_norm = readRDS(here("Data", "normalized_exprs_to_gapdh.rds")) #exprs data norm to GAPDH
drug_id_key = readRDS(here("Data", "drug_id_key.rds"))
norm_dose_resp = readRDS(here("Data", "prepped_drug_resp.rds"))
```

```{r load_signatures}
sig_files = list.files(here("Data", "Signatures"))
sigs = lapply(sig_files, function(x) readRDS(here("Data", "Signatures", x)))
names(sigs) = gsub("\\_.*", "", sig_files) #cut off everything after the first underscore

## we also need the full drug names to match with the correct drug response data. those need to be manually entered in the *same order* as the list shows
#make sure theyre spelled right! can check the `drug_id_key` table for spelling
names(sigs) #print this to make sure the order is correct
sig_drugs = c("Cisplatin", "Cytarabine", "5-Fluorouracil", "Gemcitabine", "Irinotecan",
              "Luminespib", "Paclitaxel", "Topotecan", "Vinblastine", "Vorinostat")

rm(sig_files)
```

Run the following to use 3 sigs only (what was used in the first part of the paper).

```{r 3_signatures}
sigs = sigs[c(1,3,4)]
sig_drugs=sig_drugs[c(1,3,4)]
```

To test the smaller, length-optimized sigs

```{r}
sigs[["cis"]] = c("LRRC8C", "LY6K", "MMP10","SLFN11", "STOML2","USP31","WDR3", "ZNF750")
sigs[["gem"]] = c("CRLF3", "MLKL", "POLR3G", "PROCR", "RELB", "SLFN11")
```





# Fit Dose-Response Curves

We will use the non-linear mixed-effects model to fit the dose response curves for each cell line-drug pair. The bulk of this processing will be done using the `gdscIC50` package.

## Prep Data for Model Fitting

```{r prep_dr}
prep_dose_resp <- setConcsForNlme(norm_dose_resp)
nlme_data <- prepNlmeData(prep_dose_resp, "COSMIC_ID")

rm(norm_dose_resp, prep_dose_resp) #cleanup
```

Include only the drugs we will use in the chemogram.

```{r filter_dr}
drug_ids = unique(drug_id_key %>% filter(DRUG_NAME %in% sig_drugs))

nlme_filtered = nlme_data %>% filter(DRUG_ID_lib %in% drug_ids$DRUG_ID)
```



## Fit Model & Extract Fitted Equations

The following function will use the prepped drug response data to fit an NLME models (using the `gdscIC50` package) for all cell lines and one drug at a time, calculate the median IC50 per cancer subtype, then find the observed survival in each cell line at these doses, based on what cancer subtype the cell line belongs to. The code for calculating survival at a given dose was copied from the source code for the `calcNlmeStats` function and can be found [here](https://rdrr.io/github/CancerRxGene/gdscIC50/src/R/nlme_fit_stats.R)

```{r FX_fit}
fit_nlme = function(data, drug, drug_abbrev){
  #filter to only include the data for one drug
  drug_id = drug_ids$DRUG_ID[which(drug_ids$DRUG_NAME==drug)]
  data_filt = data %>%
    filter(DRUG_ID_lib==drug_id)
  
  #fit the model (https://rdrr.io/github/CancerRxGene/gdscIC50/src/R/nlme_fit_stats.R)
  model = fitModelNlmeData(data_filt, isLargeData = F)
  stats = calcNlmeStats(model, data_filt) %>% #this returns ln(IC50) & AUC
    mutate(cell_line=as.character(CL),
           LN_IC50 = IC50,
           IC50 = exp(IC50))
  stats = left_join(stats, subtype_key, by="cell_line") #add subtype labels per cl
  
  #find median ic50s
  subtypes = unique(subtype_key$subtype)
  subtype_doses = data.frame(matrix(ncol=5, nrow=length(subtypes)))
  for(i in 1:length(subtypes)){
    temp = (stats %>% #isolate all CLs and IC50s for one subtype
      filter(subtype==subtypes[i]) %>%
      select(IC50, cell_line))
    
    subtype_doses$std_dose[i] = median(temp$IC50) #calc the standard dose we will use
    subtype_doses$subtype[i] = subtypes[i] #label subtype
    subtype_doses$sd[i] = sd(temp$IC50) #record the std dev of the ic50s
    subtype_doses$range_dif[i] = (range(temp$IC50)[2]) - (range(temp$IC50)[1]) #record range of ic50s
    subtype_doses$n[i] = length(unique(temp$cell_line)) #how many cell lines
    
    rm(temp)
  }
  subtype_doses = subtype_doses[,-c(1:5)] #rm the empty cols
  stats = left_join(stats, subtype_doses, by="subtype") #add the std doses to stats
  stats = unique(stats) #rm all the duplicates
  stats = stats %>%
    mutate(CL = cell_line, #change this name back
           dose = std_dose) #rename this to x
  
  
  #measure obs surv @ std dose per CL
  #define functions to let us predict surv at any dose, copied from source code
  logist3 <- stats::selfStart( ~ 1/(1 + exp(-(x - xmid)/scal)),
                               initial = function(mCall, LHS, data){   
                                 xy <- stats::sortedXyData(mCall[["x"]], LHS, data)
                                 if(nrow(xy) < 3) {
                                   stop("Too few distinct input values to fit a logistic")
                                 }
                                 xmid <- stats::NLSstClosestX(xy, 0.5 ) 
                                 scal <- stats::NLSstClosestX(xy, 0.75 ) - xmid
                                 value <- c(xmid, scal)
                                 names(value) <- mCall[c("xmid", "scal")]
                                 value
                               },
                               parameters = c("xmid", "scal"))
  getXfromConc <- function(xc, maxc) {
    x <- (log(xc / maxc)/log(2))+ 9
    return(x)
  }
  l3_model2 <- function(lx, maxc, xmid, scal){
    x <- getXfromConc(lx, maxc)
  #     yhat <- 1/(1 + exp(1) ^ ((x - xmid) / scal))
    yhat <- 1 - logist3(x, xmid, scal)
    return(yhat)
  }
  
  #calculate surv at standard doses
  stats = stats %>%
    mutate(surv = l3_model2(lx=dose,
                            xmid=xmid,
                            scal=scal,
                            maxc=maxc))
  
  #clean up duplicates/multiple measurements by taking mean
  all_cl = unique(stats$cell_line)
  for(i in 1:length(all_cl)){
    if(nrow(stats %>% filter(cell_line == all_cl[i])) > 1){
      stats$surv[which(stats$cell_line==all_cl[i])] = c(mean(stats$surv[which(stats$cell_line==all_cl[i])]),
                                                        mean(stats$surv[which(stats$cell_line==all_cl[i])]))
    }
  }
  
  #plot auc dist
  ggplot(stats, aes(x=auc)) +
    geom_histogram(bins=32) +
    labs(title = paste0("AUC Distribution for ", drug)) +
    theme(axis.text= element_text(size=15),
          axis.title= element_text(size=15),
          title=element_text(size=15))
  ggsave(here("Plots", "Data Prep",
              paste0(drug_abbrev, "_auc_dist.png")),
         width=7, height=7)
  
  #plot ic50 dist + standard doses
   ggplot(stats, aes(x=LN_IC50)) +
    geom_histogram(bins=32) +
     # geom_vline(xintercept=subtype_doses$std_dose[1:nrow(subtype_doses)]) +
     geom_vline(xintercept=log(stats$dose), alpha=0.7) +
     # scale_x_continuous(trans="log") +
    labs(title = paste0("IC50 Distribution for ", drug),
         x="ln(IC50)") +
    theme(axis.text= element_text(size=15),
          axis.title= element_text(size=15),
          title=element_text(size=15))
  ggsave(here("Plots", "Data Prep",
              paste0(drug_abbrev, "_ic50_dist.png")),
         width=7, height=7)


  
  stats$drug_abbrev = drug_abbrev #add in the drug abbreviation for the whole df
  
  stats = stats %>% #clean up the data to send back
     select(subtype, cell_line, drug_abbrev, surv)
  stats = unique(stats)
  
  return(stats)
}
```

Run the above function for each drug, storing the results in a list.

```{r fit}
nlme_fits = list()

#each drug only takes 10-20 seconds
for(i in 1:length(sigs)){
  print(paste(sig_drugs[i], " -- ", Sys.time()))
  nlme_fits[[i]] = fit_nlme(data=nlme_filtered, 
                            drug=sig_drugs[i], 
                            drug_abbrev=names(sigs)[i])
  names(nlme_fits)[i] = names(sigs)[i]
}

fit_results = do.call(rbind, nlme_fits) #turn the list into a data frame
rm(nlme_fits) #rm the list
```

```{r}
saveRDS(fit_results,
        here("Results",
             paste0(length(sigs), "sig_nlme_fitted_data.rds")))

fit_results = readRDS(here("Results",
             paste0(length(sigs), "sig_nlme_fitted_data.rds")))
```

## Plot Survival

Plot the survival at standardized doses as a heatmap.

```{r plot_surv}
cl_order = subtype_key[order(subtype_key$subtype), "cell_line"] #order both heatmaps by alph. subtype

plot_surv_ht = pivot_wider(fit_results, values_from="surv",
                         names_from="drug_abbrev")
plot_surv_ht = na.omit(left_join(cl_order,plot_surv_ht,by="cell_line"))
rownames(plot_surv_ht) = plot_surv_ht$cell_line

plot_surv_ht = plot_surv_ht %>%
  select(-subtype, -cell_line)

pal = colorRampPalette(c("#023e8a","#0096c7", "#48cae4", "#edede9", "#f77f00", "#d62828", "#6f1d1b"))

pheatmap::pheatmap(na.omit(plot_surv_ht)%>% 
                     select(order(colnames(plot_surv_ht))), #data to use
         cluster_cols=FALSE, cluster_rows=FALSE, #specify if rows/cols should be clustered
         angle_col=0, show_rownames = FALSE,
         color = pal(100),
         main="Survival at Standard Dose Across All Cell Lines", #title
         legend_labels = "Survival\n", #Where to save the file
         filename=here("Plots", "Data Prep", 
                       paste0(length(sig_drugs), "sig_surv_CL.png"))) #Where to save the file

rm(plot_surv_ht)
```


```{r}
ggplot(fit_results, aes(x=surv, col = drug_abbrev,)) +
  geom_density(size=1) +
  scale_color_manual(values = c("#9b2226","#f94149","#f3722c", "#f9c74f", "#90be6d",
                                "#43aa8b", "#277da1",  "#4361ee", "#7209b7", "#f72585")) +
  labs(x="Survival at Standard Dose",
       y="Frequency",
       color="Drug",
       title="Distribution of Survival at Standard Doses",
       subtitle = paste0(length(unique(fit_results$cell_line)), " Cell Lines"))
ggsave(here("Plots" ,"Data Prep", 
            paste0(length(sigs), "drug_survival_dist.png")),
       width=8, height=6)
```


# Predict Drug Response

## Generate Signature Scores

Predicting sensitivity in each cell line will be achieved using the signatures and gene expression data.

To streamline the process, we'll create a function to calculate the signature scores (median expression of the signature genes in each cell line) for one signature at time.

```{r FX_sig_score}
#F(X) to subset genes within each drug sensitivity signature, then derive the sensitivity score per cell line
sig_score = function(sig_genes, exprs, drug_abbrev){
  
  exprs_norm2 = exprs %>%
    dplyr::select(COSMIC_ID, all_of(sig_genes))
  
  #Find signature score for each cell line
  sig_exprs = exprs_norm2[,-1] %>% 
    mutate(score = apply(.,1,median), cosmic_id = exprs$COSMIC_ID)
  
  #clean up result
  score = sig_exprs %>% dplyr::select(cosmic_id, score)
  rownames(score) = sig_exprs$cosmic_id
  colnames(score) = c("cell_line", "score")
  
  score$drug_abbrev = drug_abbrev

  return(score)
}
```

The above function is now run for each of the signatures.

```{r sig_score}
#Run the above function for each sig
scores = list()

for(i in 1:length(sigs)){
  scores[[i]] = sig_score(sig_genes=sigs[[i]],
                          exprs=exprs_norm,
                          drug_abbrev=names(sigs)[i])
  names(scores)[i] = names(sigs)[i]
}
```

```{r label_score_subtypes}
#Label cancer type of each cell line
for(i in 1:length(scores)){
  #label subtypes
  scores[[i]] = full_join(scores[[i]], subtype_key, by="cell_line")

  #replace NAs with "unclassified"
  scores[[i]]$subtype = scores[[i]]$subtype %>%
  replace_na('UNCLASSIFIED') #replace any more cancer subtype NAs w/ unclassif.
}
```

```{r scores_to_df}
scores_df = do.call(rbind, scores)

scores_df_wide = pivot_wider(scores_df,
                             id_cols = c(cell_line, subtype),
                             names_from = drug_abbrev,
                             values_from = score,
                             values_fn = mean #for duplicate scores on the same cl+drug, use the mean
                             )

scores = scores_df_wide
rm(scores_df, scores_df_wide)
scores = scores %>% select(-`NA`)
```

```{r save_scores}
#Save
saveRDS(scores, here("Results",
                     paste0(length(sigs), "sig_scores.rds")))

#Reload
scores=readRDS(here("Results",
                     paste0(length(sigs), "sig_scores.rds")))

scores = na.omit(scores)

scores_long = pivot_longer(scores,
                           cols = c(names(sigs)),
                           names_to = "drug_abbrev",
                           values_to = "score")
```



## Plot Sig Scores

plot the signature scores using the same row (cell line) order as previously.

```{r plot_scores}
plot_score_ht = left_join(cl_order,scores,by="cell_line") %>% 
                     select(-cell_line, -subtype)
pal = colorRampPalette(c("darkred", "#bb3e03", "#e9d8a6", "#94d2bd", "#0a9396", "#005f73"))

pheatmap::pheatmap(na.omit(plot_score_ht), #data to use
         cluster_cols=FALSE, cluster_rows=FALSE, #specify if rows/cols should be clustered
         angle_col=0, show_rownames = FALSE,
         color = pal(100),
         main="Signature Scores Across All Cell Lines", #title
         legend_labels = "Sig Score", 
         filename=here("Plots", "Data Prep", 
                       paste0(length(sig_drugs), "sig_scores_CL.png"))) #Where to save the file

rm(plot_score_ht)
```

```{r}
ggplot(scores_long, aes(x=score, col = drug_abbrev,)) +
  geom_density(size=1) +
  scale_color_manual(values = c("#9b2226","#f94149","#f3722c", "#f9c74f", "#90be6d",
                                "#43aa8b", "#277da1",  "#4361ee", "#7209b7", "#f72585")) +
  labs(x="Sensitivity Score",
       y="Frequency",
       color="Drug",
       title="Distribution of Signature-derived Sensitivity Scores",
       subtitle = paste0(length(unique(scores_long$cell_line)), " Cell Lines"))
ggsave(here("Plots" ,"Data Prep", 
            paste0(length(sigs), "drug_sigscore_dist.png")),
       width=8, height=6)

ggplot(scores_long, aes(x=drug_abbrev, y=score)) +
  geom_violin() +
  geom_boxplot(width=0.1) +
  labs(title="Distribution of Signature-derived Sensitivity Scores",
       subtitle = paste0(length(unique(scores_long$cell_line)), " Cell Lines"))
ggsave(here("Plots" ,"Data Prep", 
            paste0(length(sigs), "drug_sigscore_box.png")),
       width=8, height=6)
```

Comparing sig scores bw different expression normalization methods

```{r}
#to run this chunk, need to run the code above once using the z-scores and another time using the gapdh-normalized scores.
#after running one option, run either line 442 or 446 (depending on which norm method matches the label), then run everything again with the other norm method. once both `comp` and `comp2` exist, continue running the rest of this chunk

# comp = scores_long %>%
#   mutate(scores_gapdh = score) %>%
#   select(-score)
# 
# comp2 = scores_long %>%
#   mutate(scores_z =score) %>%
#   select(-score)
# 
# comp3 = full_join(comp, comp2, by=c("cell_line", "drug_abbrev")) %>%
#   mutate(subtype=subtype.y) %>%
#   select(-subtype.x, -subtype.y)
# 
# ggplot(comp3, aes(x=scores_gapdh, y=scores_z)) +
#   geom_point(alpha=0.4) +
#   scale_y_continuous(limits=c(-1, 1.5), breaks=c(-1, 0, 1)) +
#   scale_x_continuous(limits=c(0.2,0.6)) +
#   labs(title = "Signature Scores From Different Expression Normalization Methods",
#        subtitle = "Normalized to GAPDH vs Z-Score Across All Genes (Within Sample)",
#        x="Sig Scores from GAPDH-normalized Expression",
#        y="Sig Scores from Expression Z-scores") +
#   facet_wrap(~drug_abbrev, labeller = labeller(drug_abbrev = 
#     c("cis" = "Cisplatin",
#       "cyta" = "Cytarabine",
#       "fu" = "5-Fluorouracil",
#       "gem" = "Gemcitabine",
#       "irino" = "Irinotecan",
#       "lum" = "Luminespib",
#       "pac" = "Paclitaxel",
#       "topo" = "Topotecan",
#       "vinb" = "Vinblastine",
#       "vor" = "Vorinostat")))
# ggsave(here("Plots" ,"Data Prep",
#             paste0(length(sigs), "drug_sig_score_correlations.png")),
#        width=8, height=7)
# 
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="cis")],
#          comp3$scores_z[which(comp3$drug_abbrev=="cis")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="cyta")],
#          comp3$scores_z[which(comp3$drug_abbrev=="cyta")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="fu")],
#          comp3$scores_z[which(comp3$drug_abbrev=="fu")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="gem")],
#          comp3$scores_z[which(comp3$drug_abbrev=="gem")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="irino")],
#          comp3$scores_z[which(comp3$drug_abbrev=="irino")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="lum")],
#          comp3$scores_z[which(comp3$drug_abbrev=="lum")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="pac")],
#          comp3$scores_z[which(comp3$drug_abbrev=="pac")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="topo")],
#          comp3$scores_z[which(comp3$drug_abbrev=="topo")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="vinb")],
#          comp3$scores_z[which(comp3$drug_abbrev=="vinb")],
#          method = 'spearman')
# cor.test(comp3$scores_gapdh[which(comp3$drug_abbrev=="vor")],
#          comp3$scores_z[which(comp3$drug_abbrev=="vor")],
#          method = 'spearman')
```





# Score Accuracy

## Generate Permutation Matrix

To quantify the accuracy of the chemogram/signature-based predictions, we calculate a unique accuracy score for each possible permutation of the drug sensitivity ranking.
The score table for 10 signatures was generated on the cluster.

```{r score_table}
n = length(sigs)
score_table = permutations(n, n, seq(n))
acc_scores=as.data.frame(score_table)
```


```{r save_perms}
#Add a column for the associated accuracy scores
m = nrow(acc_scores)
acc_scores = acc_scores %>%
  mutate(row=row_number()) %>%
  mutate(score = (m-row)/(m-1))

saveRDS(acc_scores, here("Results", paste0(n,"sig_score_table.rds")))

#The more drugs there are, the longer it takes to run the code above.
#load the matrix in if it already exists to save time
acc_scores = readRDS(here("Results", paste0(length(sigs),"sig_score_table.rds")))
```


## Reformat & Score

The following chunk was run on the cluster (and took 20 hrs to complete for 10 sigs, but less than a minute for 3 sigs).

```{r FX_score_acc}
find_score = function(data, cl){

  subset = data %>% filter(cell_line==cl) #filter to just one cl at a time

  subset = subset[order(subset$surv, decreasing = FALSE),] #put drugs in order from most response to least response
  if(nrow(subset)==length(sigs)){
    subset$true_rank = (seq(1, length(sigs))) #label the observed sensitivity rank

    subset = subset[order(subset$score, decreasing = TRUE),] #put drugs in order from most response to least response

    score = match_df(acc_scores, as.data.frame(t(subset$true_rank))) #find which score goes with the prediction ranking
    score=as.numeric(score$score)
  } else { #there may be cell lines without data for all the drugs, so those need to be filtered out here
    print("Missing drug")
    score=NA
  }

  return(score)
}
```


```{r score_acc}
all_cl = unique(fit_results$cell_line)
unscored_ranks = left_join(fit_results, scores_long,
                           by=c("cell_line", "subtype", "drug_abbrev"))

scored_ranks = data.frame(cell_line = all_cl, subtype = NA, accuracy = 0)

for(i in 1:length(all_cl)){
  print(paste(i, "out of", length(all_cl), "--", ((i/length(all_cl))*100), "%"))

  scored_ranks$cell_line[i] = all_cl[i]

  scored_ranks$subtype[i] = as.character(unique(unscored_ranks %>%
    filter(cell_line==all_cl[i]) %>%
    select(subtype)))

  value = find_score(unscored_ranks, all_cl[i])
  print(value)
  scored_ranks$accuracy[i] = value
}
```


```{r score_acc}
saveRDS(scored_ranks, here("Results", paste0(length(sigs),"sig_accuracy_scores.rds")))

scored_ranks = readRDS(here("Results", paste0(length(sigs),"sig_accuracy_scores.rds")))
```





# Plot Results

## Clean Data

Remove missing data, results for cell lines of unclassified subtypes, and format the data to be plotted.

```{r clean_scored_ranks}
#all the values with NA had missing drug response data and should be excluded
nrow(scored_ranks)
scored_ranks_full = na.omit(scored_ranks) #986 to 759 for 3 sigs, 616 after rm unclassified cell lines
                                          #968 to 670 for 10sig, 539 after rm unclass.
nrow(scored_ranks_full)

#create a new df with one row per subtype, and columns = subtype, proportion_correct
subtype = unique(scored_ranks_full$subtype)
subtype_scores = data.frame(subtype=subtype, accuracy=0, n = 0)

for (i in 1:length(subtype)){
    #Subset data per subtype
    subset = scored_ranks_full[which(scored_ranks_full$subtype == subtype[i]),]
    
    #Store subtype name
    subtype_scores$subtype[i] = subtype[i]
    
    #calc and store % of cell line survivals correctly predicted
    subtype_scores$accuracy[i] = round((sum(subset$accuracy)/nrow(subset)), digits=3)
    
    #Indicate number of cell lines in subtype (n)
    subtype_scores$n[i] = nrow(subset)
}


#order the rows by highest to lowest correct
subtype_scores = subtype_scores[order(subtype_scores$accuracy),]

#To maintain this order when we plot later, factorize the column
subtype_scores$subtype <- factor(subtype_scores$subtype,
                                      ordered = TRUE,
                                      levels = subtype_scores$subtype)



#remove unclassified cell lines
scored_ranks_full = scored_ranks_full %>% filter(subtype != "UNCLASSIFIED")

#Set order to = subtype_scores
cancer_order = (as.character(subtype_scores$subtype))[-c(which(as.character(subtype_scores$subtype)=="UNCLASSIFIED"))]

scored_ranks_full$subtype <- factor(scored_ranks_full$subtype, 
                                      ordered = TRUE, 
                                      levels = c(cancer_order))
scored_ranks_full = scored_ranks_full[order(scored_ranks_full$subtype),]

#remove unclassified from subtype score averages
subtype_scores$subtype <- factor(subtype_scores$subtype, 
                                      ordered = TRUE, 
                                      levels = c(cancer_order))
subtype_scores = subtype_scores[order(subtype_scores$subtype),]

#save
# saveRDS(scored_ranks_full, here("Results", paste0(length(sigs),"sig_accuracy_scores_cleaned.rds")))
```


## Boxplot


```{r warning=FALSE}
# beeswarm boxplot
p1 = ggplot(scored_ranks_full, aes(x = subtype, y=accuracy)) + #select data to plot
  geom_boxplot(width=0.4, fill="#9b2226", alpha = 0.6, fatten = NULL) + #boxplot, fatten=null is getting rid of the median line
  ggbeeswarm::geom_quasirandom(method="smiley", width=0.2, alpha=0.4, size = 3, color="#6a040f") + #beeswarm
  stat_summary(fun.y = mean, geom = "errorbar", aes(ymax = ..y.., ymin = ..y..), #add in a line to indicate the mean
               width = 0.4, size = 2, linetype = "solid") +
  
  geom_label(data=na.omit(subtype_scores), aes(x = subtype, y = 0, label = n), #add n
             label.padding = unit(0.15, "lines"), label.size = 0.4, size = 8) +
  
 
  scale_y_continuous(limits = c(0, 1), breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1)) +
  
  labs(title=paste0("Predictive Accuracy of ", length(sigs), "-sig Chemogram"),
       subtitle = paste0(nrow(scored_ranks_full)," Cell Lines"),
       y = "Predictive Accuracy", x = "Disease Site") +
  
  theme_bw(base_size = 15) + #theme and sizing
  theme(axis.text.y = element_text(size = 20),
        axis.text.x = element_text(angle=30, size = 20, hjust=.9),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5),
        axis.text = element_text(size = 10))

p1

ggsave(plot=p1, 
       filename=here("Plots",
                     paste0(length(sigs), "sig_accuracy_boxplots.png")), 
       width = 15, height = 9) #save plot
```


## Epithelial Only

```{r}
epi = c("HNSC", "ESCA", "BRCA",
                     "COREAD", "LIHC", #"ACC", #leave this out bc it only has one cell line
                     "STAD", "KIRC", "LUAD",
                     "LUSC","MESO", 
                     "PAAD", "THCA", "BLCA", 
                     "CESC", "UCEC", "OV", "PRAD")
```

```{r}
scored_ranks_epi = scored_ranks_full %>%
  filter(subtype %in% epi)
subtype_scores_epi = subtype_scores %>%
  filter(subtype %in% epi)

p2 = ggplot(scored_ranks_epi, aes(x = subtype, y=accuracy)) + #select data to plot
  geom_boxplot(width=0.4, fill="#9b2226", alpha = 0.6, fatten = NULL) + #boxplot, fatten=null is getting rid of the median line
  ggbeeswarm::geom_quasirandom(method="smiley", width=0.2, alpha=0.4, size = 3, color="#6a040f") + #beeswarm
  stat_summary(fun.y = mean, geom = "errorbar", aes(ymax = ..y.., ymin = ..y..), #add in a line to indicate the mean
               width = 0.4, size = 2, linetype = "solid") +
  
  geom_label(data=subtype_scores_epi, aes(x = subtype, y = 0, label = n), #add n
             label.padding = unit(0.15, "lines"), label.size = 0.4, size = 8) +
 
  scale_y_continuous(limits = c(0, 1), breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1)) +
  
  labs(title=paste0("Predictive Accuracy of ", length(sigs), "-sig Chemogram"),
       subtitle = paste0(nrow(scored_ranks_epi)," Cell Lines"),
       y = "Predictive Accuracy", x = "Disease Site") +
  
  theme_bw(base_size = 15) + #theme and sizing
  theme(axis.text.y = element_text(size = 20),
        axis.text.x = element_text(angle=30, size = 20, hjust=.9),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5),
        axis.text = element_text(size = 10))
p2

ggsave(plot=p2, 
       filename=here("Plots", "Epithelial Cancers",
                     paste0(length(sigs), "sig_accuracy_boxplots_epi.png")), 
       width = 15, height = 9) #save plot

```



## Supplemental Plots

Frequency of each ranking per drug

```{r}
#create a new df to store the rankings in
scores_sup = data.frame(matrix(nrow=0, ncol=(ncol(scores_long)+1)))
colnames(scores_sup) = c(colnames(scores_long), "rank")
#create a vector of all the cell lines to go through
cl_list = unique(scores_long$cell_line)

for(i in 1:length(cl_list)){
  #for each cell line, sort by descending sig score (most to least sensitive)
  temp = scores_long %>% filter(cell_line == cl_list[i])
  temp = temp[order(temp$score, decreasing = TRUE),]
  temp$rank = seq(1:nrow(temp)) #label the ranking from most to least sensitive
  
  scores_sup = rbind(scores_sup, temp) #combine results with the previous iteration
}
rm(temp)

scores_sup = scores_sup %>%
  filter(subtype != "UNCLASSIFIED") %>%
  mutate(rank = factor(rank, ordered=TRUE,
                       levels = seq(1:length(sigs))))
```

```{r}
#set some plotting parameters based on number of sigs used
if(length(sigs) <= 5){
  plot_rows = 1
  bar_n_displace = -10
  bar_n_angle = 0
  bar_n_size = 5
  viol_n_displace = -0.03
  
  col_values = c("#5f957d", "#406453", "#304b3e", "#203229", "#000000")[1:length(sigs)]
  fill_values = c("#c1d7cd", "#8fb7a5", "#507c68", "#48705c", "#406453")[1:length(sigs)]
  
} else {
  plot_rows = 2
  bar_n_displace = -20
  bar_n_angle = 45
  bar_n_size = 4
  viol_n_displace = -0.05
  
  col_values = c("#5f957d", "#5f957d", "#5f957d", "#406453", "#406453",
                 "#406453", "#304b3e", "#203229", "#000000", "#000000")[1:length(sigs)]
  fill_values = c("#fcfaf9", "#dae7e1", "#c1d7cd", "#a8c7b9", "#8fb7a5",
                   "#76a791","#5f957d", "#507c68", "#48705c", "#406453")[1:length(sigs)]
}

#make a funciton and list to label each facet with the full name per drug
sig_labels = list()
for(i in 1:length(sigs)){
  sig_labels[[i]] = sig_drugs[i]
  names(sig_labels)[i] = names(sigs)[i]
}
#make a function to label number of cell lines represented in each col
give.n <- function(x){
  return(c(y = median(x)*1.05, label = length(x))) 
  # experiment with the multiplier to find the perfect position
}

ggplot(scores_sup, aes(x=rank)) +
  geom_bar(position="dodge", col="#48705d", fill = "#6aa087") + 
  labs(title="Frequency of Prediction Rankings",
       x="Sensitivity Ranking", 
       y="Number of Occurrences")+

  facet_wrap(~drug_abbrev, 
             labeller=as_labeller(setNames(sig_drugs, names(sig_labels))),
             nrow=plot_rows) +
  stat_summary(fun.data = give.n, geom = "text", fun.y = median,
                  position = position_dodge(width = 0.75),
               aes(y=bar_n_displace), size = 4, angle = bar_n_angle) +
  
  theme_bw(base_size = 15) + #theme and sizing
  theme(axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5),
        axis.text = element_text(size = 10))

ggsave(here("Plots", paste0(length(sigs), "sig_prediction_freq.png")),
       width=15, height=8)
```


```{r}
ggplot(scores_sup %>% filter(subtype %in% epi), aes(x=rank)) +
  geom_bar(position="dodge", col="#48705d", fill = "#6aa087") + 
  labs(title="Frequency of Predicted Sensitivity Rankings",
       subtitle = "in Epithelial Cancers",
       x="Sensitivity Ranking", 
       y="Number of Occurrences")+

  facet_wrap(~drug_abbrev, 
             labeller=as_labeller(setNames(sig_drugs, names(sig_labels))),
             nrow=plot_rows) +
  stat_summary(fun.data = give.n, geom = "text", fun.y = median,
                  position = position_dodge(width = 0.75),
               aes(y=bar_n_displace), size = bar_n_size, angle = bar_n_angle) +
  
  theme_bw(base_size = 15) + #theme and sizing
  theme(axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5),
        axis.text = element_text(size = 10))

ggsave(here("Plots", "Epithelial Cancers",
            paste0(length(sigs), "sig_prediction_freq_epi.png")),
       width=15, height=8)
```


Drug ranking x survival distribution

```{r}
surv_sup = merge(scores_sup, fit_results, 
                 by=c("cell_line", "subtype", "drug_abbrev"))
```

```{r}
ggplot(surv_sup, aes(x=rank, y=surv, col=rank, fill=rank)) +
  geom_boxplot(position="dodge") + 
  labs(title="Distribution of Survival by Predicted Sensitivity Ranking",
       x="Sensitivity Ranking", 
       y="Survival")+
  
  scale_color_manual(values = col_values) +
  scale_fill_manual(values = fill_values) +

  facet_wrap(~drug_abbrev, 
             labeller=as_labeller(setNames(sig_drugs, names(sig_labels))),
             nrow=plot_rows) +
  stat_summary(fun.data = give.n, geom = "text", fun.y = median,
                  position = position_dodge(width = 0.75),
               aes(y=viol_n_displace), size = 5, angle = bar_n_angle) +
  
  theme_bw(base_size = 15) + #theme and sizing
  theme(axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5),
        axis.text = element_text(size = 10))

ggsave(here("Plots", paste0(length(sigs), "sig_prediction_surv.png")),
       width=15, height=8)
```

```{r}
ggplot(surv_sup %>% filter(subtype %in% epi), aes(x=rank, y=surv, fill=rank, col=rank)) +
  geom_boxplot(position="dodge") + 
  labs(title="Distribution of Survival by Predicted Sensitivity Ranking",
       subtitle = "in Epithelial Cancers",
       x="Sensitivity Ranking", 
       y="Survival")+
  
    scale_color_manual(values = col_values) +
  scale_fill_manual(values = fill_values) +

  facet_wrap(~drug_abbrev, 
             labeller=as_labeller(setNames(sig_drugs, names(sig_labels))),
             nrow=plot_rows) +
  stat_summary(fun.data = give.n, geom = "text", fun.y = median,
                  position = position_dodge(width = 0.75),
               aes(y=viol_n_displace), size = 5, angle = bar_n_angle) +
  
  theme_bw(base_size = 15) + #theme and sizing
  theme(axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5),
        axis.text = element_text(size = 10))

ggsave(here("Plots", "Epithelial Cancers",
            paste0(length(sigs), "sig_prediction_surv_epi.png")),
       width=15, height=8)
```






# Subtype Performance Correlation

correlation bw 10sig and 3 sig rankings performance among subtypes

```{r}
#all cancers
all_3 = seq(1:30)
names(all_3) = c("CLL", "ESCA", "PRAD", "MM", "LUAD",
          "BRCA","STAD","SCLC","CESC","ACC",
          "LGG","HNSC","THCA",
          "DLBC","ALL","COREAD","LUSC","PAAD",
          "OV","MESO","BLCA","SKCM","MB","UCEC",
          "NB","LIHC","LCML","KIRC","GBM","LAML")

all_10 = all_3

all_10 = all_10[order(factor(names(all_10), 
                             levels =c("CLL","ACC","MB","COREAD","ESCA","UCEC",
                                       "NB","BLCA","CESC","LIHC","THCA","PRAD",
                                       "OV","HNSC","PAAD","STAD","ALL","GBM",
                                       "LUAD","MM","BRCA","LAML","SCLC","MESO",
                                       "LUSC","LGG","LCML","DLBC","SKCM","KIRC")
                             ))]

all_3
all_10 #numbers and names should match up

cor(all_3, all_10,
    method="spearman") #0.2102336
cor(all_3, all_10,
    method="pearson") #0.2102336
cor(all_3, all_10,
    method="kendall") #0.1494253
```

```{r}
#epi
epi_3 = seq(1:17)
names(epi_3) = c("ESCA","PRAD","LUAD","BRCA",
                 "STAD","CESC","HNSC","THCA",
                 "COREAD","LUSC","PAAD","OV",
                 "MESO","BLCA","UCEC","LIHC",
                 "KIRC")

epi_10 = epi_3

epi_10 = epi_10[order(factor(names(epi_10),
                             levels =c("COREAD","ESCA", "UCEC",
                                       "BLCA","CESC", "LIHC","THCA",
                                       "PRAD","OV","HNSC","PAAD",
                                       "STAD","LUAD","BRCA","MESO","LUSC",
                                       "KIRC")
                             ))]

epi_3
epi_10

cor(epi_3, epi_10,
    method="spearman") #0.07598039
cor(epi_3, epi_10,
    method="pearson") #0.07598039
cor(epi_3, epi_10,
    method="kendall") #0.02941176
```